### ISSUE TITLE
Update docs for > 2048 token models (SuperHOT)?

### ISSUE THREAD

**Issue by tensiondriven** [johnsmith0031/alpaca_lora_4bit#129]
Hola, forgive me if this is answered elsewhere.

I'm wondering if there are any special considerations for training a 33b (or other) model that has been updated to use the positional embedding compression technique described by the SuperHOT method?

I'd like to start training my loras on the 33b llama variants that support 8K context length, but would be concerned about doing a training run if the code was artificially limiting the samples to 2048.

Are there any special considerations?  And, in the spirit of making the repo better, would it make sense to add a note to the README.md to explain this for others who might have the same question?

**Comment by johnsmith0031** [comment]
Thanks for your information! I think this idea is quite simple (although it does not reduce the VRAM cost). What we need to do is just replace RotaryEmbedding with ScaledRotaryEmbedding (not sure where the code is) and download the finetuned model (e.g. TheBloke's model). Then we can train LoRA with the same script as before.

**Comment by Jeduh** [comment]
@johnsmith0031 I've tried it and can't do anything to get the lora to perform better then llama in exllama scaled inference. Am I missing any modules? ["q_proj", "k_proj", "v_proj", "o_proj"]? I've also got all biases. Maybe I'd have to train the Rotary Encoder module for each layer? Any ideas?

**Comment by kaiokendev** [comment]
@Jeduh Can you clarify what problem you are facing? I don't think more exported modules should impact whether it works or not

**Comment by Jeduh** [comment]
If you have any time could you please check over the 2 commits I made in this [repo](https://github.com/Jeduh/Scaled-RoPE). Somethings blocking it from learning how to scale.

**Comment by kaiokendev** [comment]
It looks mostly fine to me. Is the loss not converging properly?

**Comment by Jeduh** [comment]
Thank you for taking a look :)

Well it seems to always fall bellow 1.1 at around 500 steps and converges around 1.0 avg at 2-3k steps. It looks pretty normal but no matter how much I tune the lr's decays and all that to get a lower converging curve, almost no learning gets applied in scaled inference, so weird.

**Comment by Jeduh** [comment]
![EvalLoss](https://github.com/johnsmith0031/alpaca_lora_4bit/assets/98840691/08b417e1-156d-4b0e-bb07-f0b5ac637baf)
![TrainLoss](https://github.com/johnsmith0031/alpaca_lora_4bit/assets/98840691/dfdbb03a-6d80-4737-a042-5775bded4391)

Do they look strange?

**Comment by Jeduh** [comment]
I have a hunch its due to LoRa output or intake at inference.
I remembered ExLlama always showing !! Warning: LoRA zero bias ignored which comes from this code:
`self.bias = tensors[key + ".bias"] if has_bias else None`

Seems that the 'all' bias I set, Isn't present after training? 

[This is the 2x scale LoRa adaptor](https://github.com/johnsmith0031/alpaca_lora_4bit/files/11929522/AiroborosLongFormLoRa-7b-4bit-128g.zip)

I have no idea, I'm all ears.

**Comment by kaiokendev** [comment]
> ![EvalLoss](https://user-images.githubusercontent.com/98840691/250356068-08b417e1-156d-4b0e-bb07-f0b5ac637baf.png) ![TrainLoss](https://user-images.githubusercontent.com/98840691/250356075-dfdbb03a-6d80-4737-a042-5775bded4391.png)
> 
> Do they look strange?

I'm not sure how to interpret since there is no labels, but I am assuming pink line is the loss? So it is definitely learning. As for bias, you can ignore it as this repo does not export the bias tensors by default (they will all be 0, there is a code you can change but there's no need, just set bias to none in your LoRA).

I loaded the LoRA in your post in text-generation-webui, set max_seq_len to 4096, and the compress_pos_emb to 2 and with 3271 context it was outputting fine, so I am not sure I see the same problem? Do you have up-to-date version of Exllama? I believe the CUDA allocation was causing some problems before version 0.0.4

**Comment by Jeduh** [comment]
Oh right sorry, yeah, top is eval/loss, bottom is train/loss. Ill try updating, thanks for letting me know! that makes me happy :)

**Comment by Jeduh** [comment]
@kaiokendev I've been using test_benchmark_inference.py with updated exllama but perplexity's are still in line without -ld lora. I tried it out in Ooobabooga and it does seem to work but maybe I'm just not using the right template? 

how do you test perplexities Kaioken?

**Comment by kaiokendev** [comment]
I use the perplexity code from huggingface https://huggingface.co/docs/transformers/perplexity#example-calculating-perplexity-with-gpt2-in-transformers
